api/src/main/java/de/learnlib/api/modelchecking/counterexample/Lasso.java

```diff
@@ -0,0 +1,352 @@
+/* Copyright (C) 2013-2017 TU Dortmund
+ * This file is part of LearnLib, http://www.learnlib.de/.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package de.learnlib.api.modelchecking.counterexample;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import net.automatalib.automata.concepts.InputAlphabetHolder;
+import net.automatalib.automata.concepts.Output;
+import net.automatalib.automata.fsa.DFA;
+import net.automatalib.automata.simple.SimpleAutomaton;
+import net.automatalib.automata.transout.MealyMachine;
+import net.automatalib.automata.transout.impl.compact.CompactMealyTransition;
+import net.automatalib.ts.simple.SimpleDTS;
+import net.automatalib.words.Alphabet;
+import net.automatalib.words.Word;
+import net.automatalib.words.WordBuilder;
+import net.automatalib.words.impl.Alphabets;
+
+/**
+ * A lasso is an single infinite word.
+ *
+ * The implementation is an automaton such that its singleton language is the infinite word. Also, the implementation
+ * is actually the finite representation (by unrolling the loop) of the infinite word, including information how many
+ * times the loop of the lasso is unrolled.
+ *
+ * @param <S> the state type of the automaton that contains the infinite word.
+ * @param <A> the automaton type which contains the lasso.
+ * @param <I> the input type
+ * @param <D> the output type
+ */
+@ParametersAreNonnullByDefault
+public abstract class Lasso<S, A extends SimpleDTS<S, I> & Output<I, D>, I, D>
```

Comment:
Having interfaces only is nice to have, but maybe too much work just to have an interface?
I'd say the same holds for `Query`? `Query` is an abstract class too, and could be an interface in the API.

Reply:
`Query` (and `Word` in AutomataLib) cannot be interfaces, because they declare final methods. I'm generally a fan of naming abstract classes as `AbstractXXX`, but you probably want a significant name for the concept of a `Lasso` (rather than talking about `AbstractLasso`s all the time). It was just a suggestion. If you have the same requirements (e.g. a final `hashCode`/`equals` method) or just strong preference, leave it as an abstract class. You'll probably have to add an exception to one of the checkstyle (etc.) configs though.

Reply:
I'll add the exception to checkstyle.